### ISSUE TITLE
[Bug] Infinite loop when using Pi with MTPLX as the backend provider

### ISSUE THREAD

**Issue by seelic** [youssofal/MTPLX#102]
### mtplx doctor --json
```json
{
  "diagnostics": {
    "checks": [
      {
        "command": null,
        "docs_url": "https://ml-explore.github.io/mlx/build/html/install.html",
        "expected": "macOS >= 14.0 on Apple Silicon",
        "fix": "Upgrade to macOS 14+; MLX does not support older macOS.",
        "id": "os.macos_version",
        "observed": "26.5.1",
        "severity": "error",
        "status": "pass"
      },
      {
        "command": "python3 -c \"import platform; print(platform.machine(), platform.processor())\"",
        "docs_url": "https://ml-explore.github.io/mlx/build/html/install.html",
        "expected": "native arm64 Python, not Rosetta",
        "fix": "Install/use a native arm64 Python. If needed, reinstall via Homebrew arm64 or uv.",
        "id": "python.native_arm64",
        "observed": {
          "machine": "arm64",
          "processor": "arm"
        },
        "severity": "error",
        "status": "pass"
      },
      {
        "command": null,
        "docs_url": "https://ml-explore.github.io/mlx/build/html/install.html",
        "expected": "Python >= 3.10",
        "fix": "Install Python 3.10 or newer.",
        "id": "python.version",
        "observed": "3.13.13",
        "severity": "error",
        "status": "pass"
      },
      {
        "command": "python3 -m pip install mlx",
        "docs_url": "https://ml-explore.github.io/mlx/build/html/install.html",
        "expected": "mlx importable",
        "fix": "Install MLX into this same native Python environment.",
        "id": "mlx.import",
        "observed": {
          "default_device": "Device(gpu, 0)",
          "get_active_memory": 0,
          "get_peak_memory": 0,
          "mlx": "0.31.2",
          "mlx_lm": "0.31.3"
        },
        "severity": "error",
        "status": "pass"
      },
      {
        "command": null,
        "docs_url": null,
        "expected": "estimated peak <= 80% of unified memory; 48 GiB+ recommended",
        "fix": null,
        "id": "resource.memory",
        "observed": {
          "estimated_peak_gib": 35.3,
          "unified_memory_gib": 128.0
        },
        "severity": "warning",
        "status": "pass"
      },
      {
        "command": null,
        "docs_url": null,
        "expected": "free space for model + temp download + safety headroom",
        "fix": "Free disk space or set MTPLX_MODEL_DIR to a larger volume.",
        "id": "resource.model_cache_disk",
        "observed": {
          "cache_dir": "/Users/anon/.mtplx/models",
          "free_gib": 800.66,
          "required_gib": 38.25
        },
        "severity": "warning",
        "status": "pass"
      },
      {
        "command": "mtplx pull Youssofal/Qwen3.6-27B-MTPLX-Optimized-Speed",
        "docs_url": "https://huggingface.co/Youssofal/Qwen3.6-27B-MTPLX-Optimized-Speed",
        "expected": "default model available in the HF cache or as the verified local startup model",
        "fix": "No action needed.",
        "id": "model.cache",
        "observed": {
          "hf_cache_exists": true,
          "hf_cache_path": "/Users/anon/.mtplx/models/Youssofal--Qwen3.6-27B-MTPLX-Optimized-Speed",
          "hf_cache_validation": {
            "contract_arch_id": "qwen3-next-mtp",
            "contract_error": null,
            "contract_present": true,
            "missing_files": [],
            "mtp_sidecar_candidates": [
              "mtp.safetensors",
              "mtp/weights.safetensors",
              "model-mtp.safetensors"
            ],
            "ok": true,
            "required_files": [
              "config.json",
              "tokenizer.json",
              "model.safetensors.index.json",
              "mtplx_runtime.json",
              "mtp.safetensors"
            ]
          },
          "startup_default_model": {
            "ok": true,
            "path": "/Users/anon/.mtplx/models/Youssofal--Qwen3.6-27B-MTPLX-Optimized-Speed",
            "validation": {
              "contract_arch_id": "qwen3-next-mtp",
              "contract_error": null,
              "contract_present": true,
              "missing_files": [],
              "mtp_sidecar_candidates": [
                "mtp.safetensors",
                "mtp/weights.safetensors",
                "model-mtp.safetensors"
              ],
              "ok": true,
              "required_files": [
                "config.json",
                "tokenizer.json",
                "model.safetensors.index.json",
                "mtplx_runtime.json",
                "mtp.safetensors"
              ]
            }
          }
        },
        "severity": "warning",
        "status": "pass"
      },
      {
        "command": "mtplx pull Youssofal/Qwen3.6-27B-MTPLX-Optimized-Speed",
        "docs_url": "https://huggingface.co/Youssofal/Qwen3.6-27B-MTPLX-Optimized-Speed",
        "expected": "Youssofal/Qwen3.6-27B-MTPLX-Optimized-Speed",
        "fix": "Update DEFAULT_HF_MODEL_ID to the published optimized-speed repo.",
        "id": "model.default_repo",
        "observed": "Youssofal/Qwen3.6-27B-MTPLX-Optimized-Speed",
        "severity": "error",
        "status": "pass"
      },
      {
        "command": null,
        "docs_url": "https://docs.docker.com/desktop/setup/install/mac-install/",
        "expected": "Docker Desktop installed for Open WebUI Docker path",
        "fix": "Install Docker Desktop if you want the Open WebUI Docker integration.",
        "id": "docker.binary",
        "observed": null,
        "severity": "warning",
        "status": "warn"
      },
      {
        "command": null,
        "docs_url": null,
        "expected": "port free before starting mtplx serve, or already a healthy MTPLX server",
        "fix": "Use --port 8001 or stop the existing process.",
        "id": "port.mtplx_server",
        "observed": {
          "host": "127.0.0.1",
          "open": true,
          "port": 8000
        },
        "severity": "warning",
        "status": "warn"
      },
      {
        "command": null,
        "docs_url": null,
        "expected": "port free before starting Open WebUI, or already an Open WebUI container",
        "fix": "Use a different Open WebUI host port or stop the process on 3000.",
        "id": "port.openwebui",
        "observed": {
          "host": "127.0.0.1",
          "open": false,
          "port": 3000
        },
        "severity": "warning",
        "status": "pass"
      },
      {
        "command": null,
        "docs_url": null,
        "expected": "ThermalForge or TG Pro available for explicit --max only",
        "fix": "Install ThermalForge only if you want opt-in fan boost.",
        "id": "thermal.control",
        "observed": {
          "kind": "thermalforge",
          "path": "/Users/anon/.mtplx/bin/thermalforge",
          "version": {
            "command": [
              "/Users/anon/.mtplx/bin/thermalforge",
              "--version"
            ],
            "ok": true,
            "returncode": 0,
            "stderr": "",
            "stdout": "0.1.0"
          }
        },
        "severity": "warning",
        "status": "pass"
      },
      {
        "command": null,
        "docs_url": null,
        "expected": "Low Power Mode off for best sustained decode",
        "fix": "Turn off Low Power Mode before benchmarking or serving long responses.",
        "id": "power.low_power_mode",
        "observed": {
          "available": true,
          "lowpowermode": null,
          "powermode": "0",
          "thermal": "Note: No thermal warning level has been recorded\nNote: No performance warning level has been recorded\nNote: No CPU power status has been recorded",
          "thermal_ok": true
        },
        "severity": "warning",
        "status": "pass"
      },
      {
        "command": null,
        "docs_url": null,
        "expected": "no recorded thermal or performance warning",
        "fix": "Let the Mac cool down or improve airflow before sustained benchmarks.",
        "id": "power.thermal_pressure",
        "observed": "Note: No thermal warning level has been recorded\nNote: No performance warning level has been recorded\nNote: No CPU power status has been recorded",
        "severity": "warning",
        "status": "pass"
      }
    ],
    "created_at": "2026-06-13T15:28:46+0800",
    "host": {
      "cache_dir": "/Users/anon/.mtplx/models",
      "chip": "Apple M5 Max",
      "disk_free_bytes": 859700416512,
      "disk_free_gib": 800.66,
      "mac_model": "Mac17,6",
      "machine": "arm64",
      "macos_version": "26.5.1",
      "memory_bytes": 137438953472,
      "memory_gib": 128.0,
      "platform": "macOS-26.5.1-arm64-arm-64bit-Mach-O",
      "processor": "arm",
      "python_executable": "/opt/homebrew/var/mtplx/venv-1.0.4/bin/python",
      "python_version": "3.13.13",
      "system": "Darwin"
    },
    "overall": "warn",
    "resources": {
      "default_model_size_bytes": 16430000000,
      "estimated_runtime_memory_bytes": 37904836480,
      "required_download_free_bytes": 41075000000
    },
    "schema_version": 1,
    "support_matrix": {
      "preview_test_targets": [
        "M3 Max",
        "M4 Max",
        "M3 Ultra / Mac Studio",
        "M5 Max developer machine"
      ],
      "supported": {
        "default_model": "Youssofal/Qwen3.6-27B-MTPLX-Optimized-Speed",
        "default_profile": "sustained",
        "docker": "Docker Desktop current plus previous two macOS major releases",
        "macos": ">= 14.0",
        "platform": "Apple Silicon arm64 Mac",
        "python": "native arm64 Python >= 3.10"
      }
    }
  },
  "environment": {
    "git_branch": "main",
    "git_status": "## main",
    "hf_path": "/Users/anon/.local/bin/hf",
    "mlx": {
      "default_device": "Device(gpu, 0)",
      "get_active_memory": 0,
      "get_peak_memory": 0,
      "mlx": "0.31.2",
      "mlx_lm": "0.31.3"
    },
    "platform": "macOS-26.5.1-arm64-arm-64bit-Mach-O",
    "project_root": "/Users/anon/Playground/Deo",
    "python_executable": "/opt/homebrew/var/mtplx/venv-1.0.4/bin/python",
    "python_version": "3.13.13 (main, Apr  7 2026, 18:19:01) [Clang 21.0.0 (clang-2100.0.123.102)]",
    "uv_path": null
  },
  "huggingface": {
    "cache_dir": "/Users/anon/.mtplx/models",
    "cache_exists": true,
    "cache_writable": true,
    "cached_models": 2,
    "disk_free_bytes": 859700416512,
    "disk_free_gb": 859.7,
    "token_present": true,
    "token_source": "huggingface_hub"
  },
  "policy": {
    "benchmark_exactness_smoke_context": 2048,
    "fanmax_counts_for_product_gate": false
  },
  "thermal_control": {
    "available": true,
    "clock_anchor_enabled": false,
    "clock_anchor_policy": "explicit experimental only; never used for product claims",
    "instructions": "Install ThermalForge and ensure the thermalforge CLI is on PATH.",
    "selected": {
      "kind": "thermalforge",
      "path": "/Users/anon/.mtplx/bin/thermalforge",
      "version": {
        "command": [
          "/Users/anon/.mtplx/bin/thermalforge",
          "--version"
        ],
        "ok": true,
        "returncode": 0,
        "stderr": "",
        "stdout": "0.1.0"
      }
    },
    "tools": [
      {
        "kind": "thermalforge",
        "path": "/Users/anon/.mtplx/bin/thermalforge",
        "version": {
          "command": [
            "/Users/anon/.mtplx/bin/thermalforge",
            "--version"
          ],
          "ok": true,
          "returncode": 0,
          "stderr": "",
          "stdout": "0.1.0"
        }
      }
    ]
  },
  "tools": {
    "powermetrics": "/usr/bin/powermetrics",
    "python": "/opt/homebrew/var/mtplx/venv-1.0.4/bin/python",
    "smc_atlas": null,
    "smc_atlas_exists": false,
    "sovereign": null,
    "sovereign_exists": false,
    "sudo": "/usr/bin/sudo"
  }
}
```

When using @earendil-works/pi with MTPLX as the backend provider, the agent successfully writes a backlog (epics and stories). However, when instructed to implement the first story, it enters an infinite loop, continuously repeating the same planning steps (Flyway migration creation, setting up exception handlers, setting up database configuration) without actually writing files or progressing.

```
 I'm creating the Flyway migration file to initialize the schema with a users table that includes authentication fields and constraints...                                                                                                          
                                                                                                                                                                                                                                                    
 Now I'm setting up the global exception handler to catch and format validation errors consistently, extracting field-level details and returning them with the appropriate HTTP status code.                                                       
                                                                                                                                                                                                                                                    
 Now I'm adding handlers for general exceptions and database-related errors, mapping them to appropriate HTTP status codes and error messages.                                                                                                      
                                                                                                                                                                                                                                                    
 Now I'm setting up the base configuration with the PostgreSQL connection string and other database settings.                                                                                                                                       
                                                                                                                                                                                                                                                    
 [ ... Above block repeats infinitely ... ]
```

### Exact command

mtplx start --enable-thermal-poll

### Model path or repo id

~/.mtplx/models/Youssofal--Qwen3.6-27B-MTPLX-Optimized-Quality

### Chip, RAM, macOS version

chip: Apple M5 Max, unified memory: 128.0 GB, macOS: Tahoe 26.5.1

**Comment by osenvosem** [comment]
Same with opencode and both 27b 35b models. Task execution either hangs or fall into infinite loop.

**Comment by Pflugshaupt** [comment]
Yes, I also have this happen with opencode. Even with the 27b optimized quality model - which should be best for the long contexts needed. It seems to be more likely to happen between 20k and 40k context in a session.

**Comment by codinginthebigcity-design** [comment]
same with cline

**Comment by mmmugh** [comment]
I'm having this issue with Pi code, both models, also well below the context limit. Investigating presence penalty as possible fix.

**Comment by mmmugh** [comment]
Update — I found the root cause and opened a PR (#120).

**Root cause:** MTPLX accepts `presence_penalty` / `frequency_penalty` (returns 200) but **silently drops them** — output is byte-identical with the penalty set to `0` vs `2` at temp 0. Two separate reasons: the sampler never read the fields, and the `--default-*-penalty` flags weren't plumbed through to the server. With the penalty gone, the Qwen3.6 family loses its main anti-repetition lever, so it degenerates into the repeat/"list-collapse" loops we're seeing — which fits it happening **below** the context limit (it's a sampling problem, not a context one).

It's **not** an MTP-decoding artifact: the same loop shows up on the plain target model whenever no presence penalty is applied; MTPLX just had no way to apply one.

**Fix:** #120 adds `presence_penalty` / `frequency_penalty` with OpenAI/vLLM semantics (additive on raw logits, completion-token counts, clamped `[-2,2]`), applied **per-position** inside the native-MTP verify loop so MTP output stays token-for-token identical to sequential AR decode with the same penalty. Default `0.0` is an exact no-op. The correctness guarantees (per-position ≡ AR, zero = byte-identical) are covered by unit tests in the PR.

**How to use it now** (handy for agents like Pi that don't send the field) — bake a server default:

```bash
mtplx quickstart --mtp --default-presence-penalty 1.5 …
# (and/or --default-frequency-penalty)
```

In my own testing on Qwen3.6-27B (8-bit, MTP), the penalty-off path fell into the repeat-loop while a penalty-active run ran productively without it. I'm intentionally **not** quoting a pass/fail number — I didn't get a clean controlled A/B (an unrelated hardware issue interrupted it) — but the mechanism is exactly the missing anti-repetition lever, and the PR proves the penalty now reaches the sampler. `1.0`–`2.0` is a reasonable range to try; curious whether it clears the opencode/cline cases too.

**Comment by youssofal** [comment]
Update: a concrete mitigation lever for the looping behavior has landed on main and ships in the next release.

**Presence and frequency penalties are now fully supported end-to-end** (thanks to @mmmugh's PR #120 for the sampler core, which we extended through the whole product):
- Per-request `presence_penalty` / `frequency_penalty` on `/v1/chat/completions` and `/v1/completions` are now honored — previously these fields were silently dropped, which is why setting them from Pi had no effect. If MTPLX owns generation policy for your client, opt in with the `x-mtplx-allow-client-controls: 1` header.
- Server defaults: `--default-presence-penalty` / `--default-frequency-penalty` on `mtplx serve`, `start`, and `quickstart`.
- Live-adjustable without restart via `/v1/mtplx/settings`, plus a "Presence Penalty" dial in the macOS app's inference settings, the web chat sidebar, and the dashboard.

For agent loops like the one you hit, a presence penalty around 0.5–1.0 typically breaks the repetition cycle (keep it 0 for coding tasks per Qwen's guidance — there's a new TROUBLESHOOTING section on this).

Keeping this issue open: the penalty is a lever, not a root-cause fix. The protocol-side investigation into *why* the Pi transcript drives the model into that loop (tool-history shape, stop-condition handling) is still on the list.
